Code/GraphMol/MolDraw2D/DrawMolMCHCircleAndLine.cpp

```diff
@@ -0,0 +1,287 @@
+  calcSymbolEllipse(at_idx, centre, xradius, yradius);
+  if (xradius < 1.0e-6 || yradius < 1.0e-6) {
+    return;
+  }
+  adjustLineEndForEllipse(centre, xradius, yradius, p1, p2);
+}
+
+// ****************************************************************************
+void DrawMolMCHCircleAndLine::calcSymbolEllipse(unsigned int atomIdx,
+                                                Point2D &centre,
+                                                double &xradius,
+                                                double &yradius) const {
+  centre = atCds_[atomIdx];
+  getAtomRadius(atomIdx, xradius, yradius);
+  if (drawOptions_.atomHighlightsAreCircles || !atomLabels_[atomIdx] ||
+      atomLabels_[atomIdx]->symbol_.empty()) {
+    return;
+  }
+
+  double x_min, y_min, x_max, y_max;
+  x_min = y_min = std::numeric_limits<double>::max();
+  x_max = y_max = std::numeric_limits<double>::lowest();
+  atomLabels_[atomIdx]->findExtremes(x_min, x_max, y_min, y_max);
+
+  static const double root_2 = sqrt(2.0);
+  xradius = std::max(xradius, root_2 * 0.5 * (x_max - x_min));
+  yradius = std::max(yradius, root_2 * 0.5 * (y_max - y_min));
+  centre.x = 0.5 * (x_max + x_min);
+  centre.y = 0.5 * (y_max + y_min);
+}
+
+// ****************************************************************************
+void DrawMolMCHCircleAndLine::fixHighlightJoinProblems(
+    std::vector<std::unique_ptr<DrawShape>> &atomHighlights,
+    std::vector<std::unique_ptr<DrawShape>> &bondHighlights) {
+  std::vector<unsigned int> fettledAtoms;
+  for (unsigned int i = 0; i < atomHighlights.size(); ++i) {
+    auto &atomHL = atomHighlights[i];
+    for (auto &bondHL : bondHighlights) {
+      if (atomHL->atom1_ == bondHL->atom1_ ||
+          atomHL->atom1_ == bondHL->atom2_) {
+        // A multicoloured bond highlight is a polyline, with 4 points, so 2 is
+        // opposite 0.  Normally a bond highlight is a single line, with
+        // only 2 points.
+        int p2 = bondHL->points_.size() == 2 ? 1 : 2;
+        bool ins = doesLineIntersectArc(
+            atomHL->points_[0], atomHL->points_[1].x, atomHL->points_[1].y, 0.0,
+            360.0, 0.0, bondHL->points_[0], bondHL->points_[p2]);
+        if (!ins) {
+          fettledAtoms.push_back(i);
+          while (!ins) {
+            double dist1 = (atomHL->points_[0] - bondHL->points_[0]).length();
+            double dist2 = (atomHL->points_[0] - bondHL->points_[p2]).length();
+            double maxrad =
+                std::max(atomHL->points_[1].x, atomHL->points_[1].y);
+            // If it's a filled highlight, the bond highlight won't have been
+            // truncated at the outer radius of the circle, because there's no
+            // need.  Set the dist2 to the radius now.
+            if (dist2 < 1.0e-4) {
+              dist2 = maxrad;
+            }
+            double upscaler = 1.25 * std::min(dist1, dist2) / maxrad;
+            if (upscaler <= 1.0) {
+              upscaler = 1.1;
+            }
```

Comment:
There's a fair amount of magic in here.
Would be nice to add a comment explaining what it means

Reply:
It doesn't make the comment any less valid, but this is the old, original code which you've already reviewed once :-).  Perhaps I should have been clearer in the PR notes - the CircleAndLine class was just the old DrawMolMCH class, pretty much.  I will try and work out what's going on and put some explanation in.

Reply:
> It doesn't make the comment any less valid, but this is the old, original code which you've already reviewed once :-). 

I reserve the right to attempt to correct my past mistakes! :-)